tortoise/threshold.go

```diff
@@ -84,16 +85,52 @@ func computeExpectedWeight(weights map[types.EpochID]weight, from, to types.Laye
 
 func computeLocalThreshold(config Config, epochWeight map[types.EpochID]weight, lid types.LayerID) weight {
 	threshold := weightFromUint64(0)
-	// TODO(dshulyak) expected weight for local threshold should be based on last layer.
 	threshold = threshold.add(epochWeight[lid.GetEpoch()])
 	threshold = threshold.fraction(config.LocalThreshold)
 	return threshold
 }
 
-func computeGlobalThreshold(config Config, epochWeight map[types.EpochID]weight, target, last types.LayerID) weight {
+func computeThresholdForLayers(config Config, epochWeight map[types.EpochID]weight, target, last types.LayerID) weight {
 	expected := computeExpectedWeight(epochWeight, target, last)
 	threshold := weightFromUint64(0)
 	threshold = threshold.add(expected)
 	threshold = threshold.fraction(config.GlobalThreshold)
 	return threshold
 }
+
+func getConfidenceWindow(config Config, state *commonState, tmode mode) types.LayerID {
+	target := state.verified.Add(1)
+	window := state.last
+	if tmode.isFull() && state.last.Difference(target) > config.FullModeRerunWindow {
+		window = target.Add(config.FullModeRerunWindow)
+	} else if tmode.isVerifying() && state.last.Difference(target) > config.VerifyingModeRerunWindow {
+		window = target.Add(config.VerifyingModeRerunWindow)
+	}
+	return window
+}
+
+// updateThresholds recomputes local and global thresholds:
+// - when last layer is updated
+// - when verified layer is updated
+// - when switching from one mode into the other.
+func updateThresholds(logger log.Log, config Config, state *commonState, tmode mode) {
+	state.localThreshold = computeLocalThreshold(config, state.epochWeight, state.last)
+
+	window := getConfidenceWindow(config, state, tmode)
+	if window.Before(state.processed) {
+		window = state.processed
```

Comment:
why should window be reset to state.processed?

and if this is necessary, why not do it inside `getConfidenceWindow`

Reply:
so let's say window is 100, we want to verify layer 1 and the processed layer is 400. it means that we counted votes from layers 2 up to 400, so expected weight should be no less than that

Reply:
it will the case if we can't decide on the layer

Reply:
i am also using getConfidenceWindow(config, state, tmode) to decide if i need to recompute thresholds